memory protection - x86 userspace implementation

tyrantlizard77 opened the pull request:
This patch series defines the functions that architectures must implement to support user-level threads.

Per thread permission tracking for kernel objects is also implemented, filling in some stub functions in userspace.c

An implementation of userspace APIs is provided for x86 targets.

Commit:
_thread_entry: don't use _current

Thread may be in user mode when it returns and can't look at
_current. Use k_current_get() which will be a system call.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: specify arch functions for userspace

Any arches that support userspace will need to implement these
functions.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: implement userspace APIs

- _arch_user_mode_enter() implemented
- _arch_is_user_context() implemented
- _new_thread() will honor K_USER option if passed in
- System call triggering macros implemented
- _thread_entry_wrapper moved and now looks for the next function to
call in EDI

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: add K_USER flag and _is_thread_user()

Indicates that the thread is configured to run in user mode.
Delete stub function in userspace.c

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
userspace: add thread-level permission tracking

Now creating a thread will assign it a unique, monotonically increasing
id which is used to reference the permission bitfield in the kernel
object metadata.

Stub functions in userspace.c now implemented.

_new_thread is now wrapped in a common function with pre- and post-
architecture thread initialization tasks.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: add common functions for user mode

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andyross commented on arch/x86/core/thread.c:
```diff
@@ -188,3 +143,95 @@ void _new_thread(struct k_thread *thread, k_thread_stack_t stack,
+	__asm__ volatile ("mov %%cs, %[cs_val]" : [cs_val] "=r" (cs));
+
+	/* Mask all but the lowest 2 bits, which is the CPL information */
+	cs &= 0x3;
```
FWIW: you could also just define a single "kernel" code segment and test for that descriptor exactly instead of masking off the ring level.  This version is more correct, but takes a few more instructions and is likely to be on a hot path.

Related point: seems like this really wants to be an inline function.  The calling overhead is significantly bigger than the time to execute the meat.

andrewboie replied:
good idea on both counts, will do

andyross commented on arch/x86/core/thread.c:
```diff
@@ -188,3 +143,95 @@ void _new_thread(struct k_thread *thread, k_thread_stack_t stack,
+			    MMU_ENTRY_SUPERVISOR),
+			   (MMU_PTE_P_MASK | MMU_PTE_RW_MASK |
+			    MMU_PTE_US_MASK));
+
```
Same point as above: does a far call that changes the ring level do a TLB flush?  I'm not sure that it does.  Normal OSes will always rewrite CR3 on context switch, which will obviously flush the cache.  But this code looks like it's just changing page tables in place but not actually telling the CPU that the page tables have been changed.

andrewboie replied:
_x86_mmu_set_flags() flushes pages for the relevant memory addresses.

AdithyaBaglody replied:
Was looking into this today. Actually we can think of this if it improves the performance of our mmu. Instead of calling a cache flush for each page write we call it only once. So what we do is, just rewrite the same start address into CR3 when we finish all the page table manipulations.

andrewboie replied:
I considered this, but it's not clear what the tradeoffs are.
Resetting CR3 is just one instruction, but the CPU has to do a lot of work once that is done since it has to invalidate everything.
It's not clear what the threshold is, where resetting CR3 performs better than using "invlpg" on only those pages which were changed.

Some discussion here: https://stackoverflow.com/questions/28384234/when-to-do-or-not-do-invlpg-mov-to-cr3-to-minimize-tlb-flushing

andyross commented on arch/x86/core/userspace.S:
```diff
@@ -0,0 +1,145 @@
+SECTION_FUNC(TEXT, _x86_syscall_entry_stub)
+	push	%esi		/* arg5 */
+	push	%edi		/* arg4 */
+	push	%ebx		/* arg3 */
```
Sort of annoying that you need to push stuff on the stack for arguments which are only going to be valid for a small minority of syscalls.  Maybe more cycle-counting than it's worth, but it wouldn't be *that* much more work to have a separate syscall interrupt registered for each argument count.

andrewboie replied:
gonna punt on this for now, but maybe revisit in the future. seems like other OSes use a single vector for system calls.

andyross commented on arch/x86/core/userspace.S:
```diff
@@ -0,0 +1,145 @@
+	call	_x86_syscall_entry
+
+	/* EAX now contains return value. Pop or xor everything else to prevent
+	 * information leak from kernel mode.
```
Started typing a comment here that popping the registers was likely a clumsy way to do this, but never mind, I'm wrong.  These are all 1-byte instructions (and they save the ESP adjustment too), and the extra memory work is essentially free on iamcu, and even on bigcore it's going to be cache-resident anyway.  Looks weird, but really clever.  Cool.

andyross commented on arch/x86/core/userspace.S:
```diff
@@ -0,0 +1,145 @@
+ * A one-way trip to userspace.
+ */
+SECTION_FUNC(TEXT, _x86_userspace_enter)
+	add	$0x4, %esp /* Discard return address on stack */
```
Hah, here's a spot where the reverse is true: make this a POP instruction instead (looks like %esi is going to be clobbered immediately below and is a good choice).   Much smaller instruction than the add, which the assembler (probably) won't be able to optimize because of the side effect from POP (it's smart, but not that smart).

andrewboie replied:
good idea

andyross commented on arch/x86/core/userspace.S:
```diff
@@ -0,0 +1,145 @@
+	 *
+	 * For security reasons, we must erase the entire user stack.
+	 * We don't know what previous contexts it was used and do not
+	 * want to leak any information.
```
Not loving this.  Obviously yes: in the abstract a thread stack might have anything in it and we need to start from a known state.  But the overwhelming majority of these stacks are going to be static things whose stacks where started from kernel mode at system startup and which haven't been presented to anything but OS code yet.  There ought to be a way to capture that info somewhere to avoid the double-memset.

Maybe something that would be better captured by a jira to track the needed optimization.

Also: why is this being done in assembly?  Can't a precondition for calling this simply be "the stack must be zero" and so whatever kernel code is responsible for doing that does its own memset() in C-land?  It's not like it's complicated code, but as we grow this is going to have to be rewritten (in assembly!) for every architecture.

andrewboie replied:
The stack we are zeroing is for the currently running thread, if you memset while on that stack you will explode. at minimum, you would have to switch to the kernel stack first, then call memset. but perhaps I can make a call to memset here. let me try this.

> But the overwhelming majority of these stacks are going to be static things whose stacks where started from kernel mode at system startup and which haven't been presented to anything but OS code yet

I was thinking we would implement everything with extreme paranoia for the first pass, and then later on either introduce kconfig options or runtime flags to relax things.

andrewboie replied:
the stack has to be zeroed here, as you cannot zero out a stack you are currently on,
I played around with calling memset from here, but it's soooo many more instructions than doing a 'rep stos'. if other arches want to call memset they can, as long as they switch to the kernel stack first. i'd like to leave this alone on x86.

andyross commented on arch/x86/core/userspace.S:
```diff
@@ -0,0 +1,145 @@
+	sub	%edi, %ecx
+
+#ifdef CONFIG_INIT_STACKS
+	mov	$0xAAAAAAAA, %eax
```
Another good reason not to duplicate the stack clear from C is that you need duplicate handling for all the somewhat obscure features.

andrewboie replied:
you've confused me here.
there is no C code specifically for clearing stacks.

andyross commented on arch/x86/core/userspace.S:
```diff
@@ -0,0 +1,145 @@
+	mov	%bx, %ds
+	mov	%bx, %es
+	mov	%bx, %fs
+	mov	%bx, %gs
```
Made this comment in a previous patch set, but just to repeat: I really do not like the use of the same segment descriptor for ES/FS/GS.  We want those segments to generate *exceptions* when used, so that crazy user code doesn't get the idea that they're supposed to work just like the data segment.

In other OSes, these segments are used for alternate state, like TLS.  Given this setup, making that work in the future requires potentially breaking usermode code.

andrewboie replied:
Some instructions deliberately do not use DS. The "rep stosl" used earlier in this very same function looks at memory in ES:EDI!!

http://x86.renejeschke.de/html/file_module_x86_id_306.html

I get crashes if I take your advice here, I prefer to leave it.

andyross commented on arch/x86/core/userspace.S:
```diff
@@ -0,0 +1,145 @@
+	push	$_thread_entry	/* EIP */
+
+	/* We will land in _thread_entry() in user mode after this */
+	iret
```
This seems a little clumsy, having to fake an interrupt stack frame.  Can't you just do a far call here?  Must the initial entry really be an IRET?

andrewboie replied:
to make a long story short, no. it has to be an iret.

andyross commented on arch/x86/include/kernel_arch_data.h:
```diff
@@ -45,6 +45,8 @@
 #define MAIN_TSS	0x18
 #define DF_TSS		0x20
+#define USER_CODE_SEG	0x2b /* at dpl=3 */
+#define USER_DATA_SEG	0x33 /* at dpl=3 */
```
Just curious: why these particular descriptor numbers?  The others are all multiples of 8, but these look funny.

andrewboie replied:
the two low bits are the privilege level, which is why 0x3 is ANDed in.

andyross commented on include/arch/x86/arch.h:
```diff
@@ -530,12 +531,112 @@ extern FUNC_NORETURN void _SysFatalErrorHandler(unsigned int reason,
+			 : "=a" (ret)
+			 : "a" (call_id), "d" (arg1), "c" (arg2),
+			   "b" (arg3), "D" (arg4), "S" (arg5),
+			   [vector] "i" (CONFIG_X86_SYSCALL_VECTOR));
```
Style: so, symbolic constants are always good, but not a whole lot of people understand the GCC syntax for named replacement arguments or even the fact that the arguments can be things other than registers.

I honestly think this would be a lot more readable and perfectly maintainable as "int 0x80", honestly.

andrewboie replied:
this function also marshals all the arguments in the correct registers. please look at the generated assembly. a bare "int 0x80" won't work, and I'm not comfortable with hard-coding the vector id.
if people don't understand GCC extended inline ASM syntax, they can google it.

andyross replied:
Of course "int 0x80" would work, that's all this is doing.  That directive says replace the named expression "vector" with the integer-typed expression CONFIG_X86_SYSCALL_VECTOR.  Then it hands that expanded string to the assembler which sees "int 0x80" (or more likely "int 128" I guess).

Obviously not a big deal.  I just don't really see this as all that readable given something (the syscall interrupt vector) that is a static "known to everyone and part of the well-documented calling conventions" expression.

andrewboie replied:
@andyross no, it's not just doing "int 0x80"

Please review https://gcc.gnu.org/onlinedocs/gcc/Machine-Constraints.html#Machine-Constraints

The arguments have to all go in very specific registers. It happens to be (by design) that call_id, arg1, arg2 will already be in the right registers for the IAMCU calling convention, but for SYS V they have to be moved, as well as any args past arg2.

andyross replied:
I'm not explaining myself well enough:  I totally get what you're doing with the **register** constraints here and it makes sense.  I'm just saying that you're jumping through hoops to insert "0x80" as a soft-configured string using a slightly obscure GCC feature and a ton of boilerplate, when realistically all you need to do is document "0x80 is the syscall trap" somewhere and use a 0x80 literal instead of the [vector] and Kconfig stuff.

andrewboie replied:
i think the point you are actually trying to argue is that "we don't need to allow the end user to configure what vector to use for system calls", is that a fair assessment?

andrewboie replied:
i think my thought process was that the IDT might be configured to have less than 256 entries, but if we're dealing with 4K pages I can't imagine people would be keen on optimizing the IDT size

andyross replied:
Exactly that.  Or if you really want to, do the parametrized INT instruction with a preprocessor##paste or some other generator and not the assembly trick.

andyross commented on include/arch/x86/arch.h:
```diff
@@ -530,12 +531,112 @@ extern FUNC_NORETURN void _SysFatalErrorHandler(unsigned int reason,
+}
+
+static inline u32_t _arch_syscall_invoke4(u32_t call_id, u32_t arg1, u32_t arg2,
+					  u32_t arg3, u32_t arg4)
```
Another efficiency point: the C side of this marshalls the inbound function by putting an extra argument on the **front** of the list.  So the original (iamcu) function took "call_id" in EAX, but now needs to write it to EDX needlessly.  If you put call_id at the end (and fix up the entry code on the kernel side accordingly), the compiler can see that the arguments can stay in the registers they're already in.

andrewboie replied:
I'm pretty sure call_id stays in EAX...double checking thus

andrewboie replied:
yes it stays in EAX.
"a" (call_id) means the argument goes in EAX.

https://gcc.gnu.org/onlinedocs/gcc/Machine-Constraints.html#Machine-Constraints

andyross replied:
Sorry, I wrote that complaint wrong: the outer function would have had **arg1** in EAX, but we need to move it to EDX, right?

andrewboie replied:
no we don't. why do you think it needs to be moved to EDX? I'm very confused.
_x86_syscall_entry_stub expects call_id in EAX.
Nothing gets moved.

andrewboie replied:
you seem to be confusing call_id and arg1

andyross replied:
Maybe I'm misunderstanding the intent of the API?  I'm assuming you've got some call that looks like:

      int k_my_call(int my_arg);

And are wrapping it with something that looks like:

    int k_my_call(int my_arg)
    {
        _arch_syscall_invoke1(MY_SYSCALL_NUM, my_arg);
    }

In this case you'd arrive in k_my_call with my_arg in EAX, which the compiler will then need to move to EDX.  If MY_SYSCALL_NUM was the *last* argument, you wouldn't have this problem and the compiler would leave EAX unmolested.

andrewboie replied:
aha, *now* I see it. OK let me give that a try, thanks for bearing with me

lpereira replied:
Would the cost of having an indirect call be too high?

```C
static const void (*syscall_table)()[] = {
    [SYS_FOO] = _sys_foo,
    [SYS_BAR] = _sys_bar,
    [SYS_BAZ] = _sys_baz,
};

static inline long _arch_syscall_invoke1(int num, u32_t arg)
{
    return syscall_table[num](arg);
}

static inline long _arch_syscall_invoke2(int num, u32_t arg1, u32_t arg2)
{
    return syscall_table[num](arg1, arg2);
}
```

(Error checking omitted.)

These being inline will most likely translate to a `lea` + `jmp *` if the compiler is smart enough to figure out this is actually a tail call. (This usually happens on x86, but not on xtensa.)

lpereira replied:
Maybe I've made the comment in the wrong place -- yeah, this code wouldn't run in user land. It would be inside the interrupt trap, and `num` would have to be validated to avoid userland calling whatever it wants inside the kernel.

andrewboie replied:
i don't see a good way to validate the function pointer without adding it to the whole gperf contraption, which would waste memory since we wouldn't need a struct _k_object associated with it. i think it's simpler to just have a (probably generated table) mapping call IDs to function pointers.

lpereira replied:
You don't need to use gperf here: that's precisely what's the `syscall_table` thing in the thing I wrote is: an array mapping the system call number to a function pointer. It's easy to validate the number: just check the bounds.

andrewboie replied:
we're in violent agreement then :)

andyross commented on kernel/thread.c:
```diff
@@ -523,8 +523,21 @@ void _k_thread_group_leave(u32_t groups, struct k_thread *thread)
 }
-#endif
+
+int _is_user_context(void)
```
Same comment as on the wrapper: seems like this really wants to be an inline so that code that needs to use it (which is really likely to be on a hot path) can get the one-instruction version instead of the double-wrapped-function-call.

andrewboie replied:
agreed

andyross reviewed (changes requested):
Misc. notes and whing, most not-super-important.

The -1 is because I'm deathly afraid we're missing a TLB flush on the privilege transition.  This would be a silent security failure on our setup where the addresses are untranslated, which would be really bad.  I'm not **sure** that there isn't some handling elsewhere or a quirk of the MMU that does the flush for us, but someone needs to show me where it is (and we should probably write a test to check for this stuff). :)

andrewboie commented on arch/x86/core/thread.c:
```diff
@@ -188,3 +143,95 @@ void _new_thread(struct k_thread *thread, k_thread_stack_t stack,
+	 * of the system call or driver API
+	 */
+	printk("syscall stub %u(%u, %u, %u, %u, %u)\n", call_id, arg1, arg2,
+	       arg3, arg4, arg5);
```
sure. actually the stub should be in common kernel code too, this landing site should be arch-agnostic.

andrewboie commented on arch/x86/core/thread.c:
```diff
@@ -188,3 +143,95 @@ void _new_thread(struct k_thread *thread, k_thread_stack_t stack,
+	_x86_mmu_set_flags((void *)incoming->stack_info.start,
+			   ROUND_UP(incoming->stack_info.size, MMU_PAGE_SIZE),
+			   MMU_ENTRY_USER, MMU_PTE_US_MASK);
+
```
_x86_mmu_set_flags() does the TLB flush. it calls invlpg for each page that was modified.

andrewboie commented on arch/x86/core/swap.S:
```diff
@@ -369,3 +392,81 @@ skipIntLatencyStop:
+	|| defined(CONFIG_X86_IAMCU)
+/**
+ *
+ * @brief Adjust stack/parameters before invoking thread entry function
```
i changed it too. it now looks for the entry point in EDI, before it used to be hard-coded. this is so we can have it call _thread_entry or _arch_user_mode_enter depending on whether K_USER was passed to k_thread_create()

andrewboie commented on kernel/include/nano_internal.h:
```diff
@@ -49,6 +49,12 @@ extern void _new_thread(struct k_thread *thread, k_thread_stack_t pStack,
 
+extern void _setup_new_thread(struct k_thread *new_thread,
+			      k_thread_stack_t stack, size_t stack_size,
+			      void (*entry)(void *, void *, void*),
```
actually I think there is, just infrequently used. I'll certainly use it in these functions I am adding.

andrewboie commented on kernel/thread.c:
```diff
@@ -248,18 +264,33 @@ static void schedule_new_thread(struct k_thread *thread, s32_t delay)
+		       k_thread_stack_t stack, size_t stack_size,
+		       void (*entry)(void *, void *, void*),
+		       void *p1, void *p2, void *p3,
+		       int prio, u32_t options)
```
_new_thread() is arch-specific, whereas _setup_new_thread() is arch-independent.
Nobody should ever call _new_thread() directly.
I can take another pass at the documentation.

andrewboie commented on kernel/thread.c:
```diff
@@ -248,18 +264,33 @@ static void schedule_new_thread(struct k_thread *thread, s32_t delay)
 	__ASSERT(!_is_in_isr(), "Threads may not be created in ISRs");
-	_new_thread(new_thread, stack, stack_size, entry, p1, p2, p3,
-		    prio, options);
-	_k_object_init(new_thread);
```
yeah. object validation is now part of CONFIG_USERSPACE instead of being a separate thing.

andrewboie commented on arch/x86/core/swap.S:
```diff
@@ -144,8 +149,26 @@ SECTION_FUNC(TEXT, __swap)
+
+#ifdef CONFIG_X86_IAMCU
+	push	%eax
+	mov	%ecx, %edx
```
I'll try swapping the use of ECX/EDX, seems simplest

Commit:
kernel: add _k_syscall_entry stub

This is the kernel-side landing site for system calls. It's currently
just a stub.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: reorganize CONFIG_USERSPACE

This now depends on a capability Kconfig.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
Updated changes:
- Some hot-path APIs changed to inline functions
- All APIs which could directly or indirectly called by user code now defined in kernel.h instead of nano_internal.h
- Syscall macros now have the call ID as the last argument, to minimize moving arguments around
- FS/GS registers no longer set
- system call vector now fixed at 0x80h
- some adds in assembly changed to pops to minimize code size
- optimized register use in _Swap()
- system call stub now in common kernel code
- use k_thread_entry_t where appropriate

andrewboie commented:
@andyross just in case you didn't see my now-hidden comment, _x86_mmu_set_flags() invalidates page table entries as appropriate

andyross reviewed (approved):
All replies made sense to me.  Looks good.
